rust/lance/src/datatypes/schema.rs

```diff
@@ -96,15 +150,27 @@ impl Schema {
         })
     }
 
-    pub fn project_by_ids(&self, column_ids: &[i32]) -> Result<Self> {
-        let protos: Vec<pb::Field> = self.into();
+    /// Iterates over the fields using a pre-order traversal
+    ///
+    /// This is a DFS traversal where the parent is visited
+    /// before its children
+    fn fields_pre_order(&self) -> SchemaFieldIterPreOrder {
+        SchemaFieldIterPreOrder::new(self)
+    }
 
-        let filtered_protos: Vec<pb::Field> = protos
+    /// Returns a new schema that only contains the fields in `column_ids`.
+    ///
+    /// This projection can filter out both top-level and nested fields
+    pub fn project_by_ids(&self, column_ids: &[i32]) -> Self {
+        let filtered_fields = self
+            .fields
             .iter()
-            .filter(|p| column_ids.contains(&p.id))
-            .cloned()
+            .filter_map(|f| f.project_by_filter(&|f| column_ids.contains(&f.id)))
```

Comment:
i forget. does it return the path from the root node to deep nested field if only the leave fields are specified via `column_ids` ?

Reply:
Good catch.  I added a test case for this and fixed it.  In order to maintain the old behavior we need to keep parent fields if they have any matching children.  I think this is the most intuitive behavior as well.  We can always add support for collapsing a schema (e.g. remove structs with only one child) later as a different method.